logger/logger.go

```diff
@@ -0,0 +1,47 @@
+package logger
+
+import (
+	"io"
+	"log"
+	"os"
+
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+)
+
+// NewLogger creates a new logger that can write to disk.
+//
+// The function also returns a closer function that should be called when we
+// stop using the logger, typically deferred in main.
+func NewLogger(level, logfile string) (logger *logrus.Logger, closer func(), err error) {
+	logger = logrus.New()
+	// Parse and set log level.
+	logLevel, err := logrus.ParseLevel(level)
+	if err != nil {
+		return nil, nil, errors.AddContext(err, "invalid log level: "+level)
+	}
+	logger.SetLevel(logLevel)
+	// Open and start writing to the log file, unless we have an empty string,
+	// which signifies "don't log to disk".
+	if logfile != "" {
+		var fh *os.File
+		fh, err = os.OpenFile(logfile, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644)
+		if err != nil {
+			return nil, nil, errors.AddContext(err, "failed to open log file")
+		}
+		logger.SetOutput(io.MultiWriter(os.Stdout, fh))
+		// Create a closer function which flushes the content to disk and closes
+		// the log file gracefully.
+		closer = func() {
```

Comment:
The more common way to do this would be passing the file in and handling the defer outside of this function and passing it in as an `io.Writer` as I mentioned in the other PR already (to also get rid of the sentinel value) or to wrap the logger in your own type and implement `Close` on it which you then defer in the outer scope.
That avoids allocating the function on the heap.

Reply:
Implemented a wrapper type.